buffer/src/main/java/io/netty/buffer/search/ShiftingBitMaskSearchProcessor.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+ package io.netty.buffer.search;
+
+/**
+ * Implements Shifting Bit Mask string search algorithm as {@link io.netty.util.ByteProcessor}.
+ * @see SearchProcessorFactory
+ */
+public class ShiftingBitMaskSearchProcessor implements SearchProcessor {
+
+    private final long[] bitMasks;
+    private final long successBit;
+    private long currentMask;
+
+    ShiftingBitMaskSearchProcessor(long[] bitMasks, long successBit) {
+        this.bitMasks = bitMasks;
+        this.successBit = successBit;
+    }
+
+    @Override
+    public boolean process(byte value) {
+        currentMask = ((currentMask << 1) | 1) & bitMasks[value & 0xff];
```

Comment:
have you tried inspecting ASM here? Look if there is any `PlatformDependent` method that can save bound checks there.

Reply:
@franz1981 I've checked the assembly, also tried `Byte.toUnsignedInt` (which I think intrinsics to `movzbl`), nothing has helped to eliminate the bounds check. JDK 8 / 11 / 13 / GraalVM.
Looks like `PlatformDependent.getByte` is something worth trying, thanks for the tip - will update later.

Reply:
So, adding this to `PlatformDependent0`:
```java
    static long getLong(long[] data, int index) {
        return UNSAFE.getLong(data, LONG_ARRAY_BASE_OFFSET + LONG_ARRAY_INDEX_SCALE * index);
    }
```
and using it in `ShiftingBitMaskSearchProcessor` gives 2-8% improvement.

Assembly with **Array** (`process` method is inlined and 4x loop-unrolled, so I'm only pasting one unroll piece here):
```
  2.36%  │  │ ││  │      0x000000010a0c791e: movzx  r9d,BYTE PTR [rbx+rcx*1+0x12]
         │  │ ││  │                                                    ;*iand
         │  │ ││  │                                                    ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@17 (line 37)
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ ││  │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  0.70%  │  │ ││  │      0x000000010a0c7924: mov    rax,rdx
  2.69%  │  │ ││  │      0x000000010a0c7927: shl    rax,1
  0.78%  │  │ ││  │      0x000000010a0c792a: mov    r8d,r10d
  1.85%  │  │ ││  │      0x000000010a0c792d: add    r8d,0x2            ;*iinc
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@21 (line 1335)
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ ││  │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  0.91%  │  │ ││  │      0x000000010a0c7931: or     rax,0x1            ;*lor  ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@8 (line 37)
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │  │ ││  │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ ││  │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  2.84%  │  │ ││  │      0x000000010a0c7935: cmp    r9d,esi
         │  │ ││╭ │      0x000000010a0c7938: jae    0x000000010a0c79ef  ;*laload
         │  │ │││ │                                                    ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@18 (line 37)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ │││ │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  0.53%  │  │ │││ │      0x000000010a0c793e: and    rax,QWORD PTR [rbp+r9*8+0x10]
         │  │ │││ │                                                    ;*land
         │  │ │││ │                                                    ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@19 (line 37)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ │││ │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  2.55%  │  │ │││ │      0x000000010a0c7943: mov    r9,rax
  0.81%  │  │ │││ │      0x000000010a0c7946: and    r9,r13             ;*land
         │  │ │││ │                                                    ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@31 (line 38)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │  │ │││ │                                                    ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │  │ │││ │                                                    ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  3.31%  │  │ │││ │      0x000000010a0c7949: test   r9,r9
         │  │ │││ │      0x000000010a0c794c: jne    0x000000010a0c7a43  ;*ifne
```
(bound check is `cmp r9d,esi; jae`)

Assembly with **Unsafe** (again, 1/4 piece of loop unroll):
```
  6.95%  │     │    0x000000010c2cc973: shl    rbx,1
  0.33%  │     │    0x000000010c2cc976: movsxd rax,r10d
  0.26%  │     │    0x000000010c2cc979: movzx  esi,BYTE PTR [rdi+rax*1+0x11]
         │     │                                                  ;*iand
         │     │                                                  ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@17 (line 36)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │     │                                                  ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  0.11%  │     │    0x000000010c2cc97e: or     rbx,0x1
  6.88%  │     │    0x000000010c2cc982: and    rbx,QWORD PTR [rdx+rsi*8+0x10]
         │     │                                                  ;*land
         │     │                                                  ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@21 (line 36)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │     │                                                  ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  1.56%  │     │    0x000000010c2cc987: mov    rsi,rbx
  0.42%  │     │    0x000000010c2cc98a: and    rsi,r8             ;*land
         │     │                                                  ; - io.netty.buffer.search.ShiftingBitMaskSearchProcessor::process@33 (line 38)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByteAsc0@11 (line 1336)
         │     │                                                  ; - io.netty.buffer.AbstractByteBuf::forEachByte@14 (line 1316)
         │     │                                                  ; - io.netty.microbench.search.SearchBenchmark::shiftingBitMask@11 (line 197)
  6.74%  │     │    0x000000010c2cc98d: test   rsi,rsi
         │   ╭ │    0x000000010c2cc990: jne    0x000000010c2cca19  ;*ifne
```
(which looks really good to me, except for maybe some instruction ordering)

Of course, same change could be done for KMP and Aho-Corasic (requiring to add `getInt` and `getObject` to `PlatformDependent`).

Reply:
Tried `PlatformDependent` with all algos. Up to 8% improvement for Shifting bit mask, up to +21% for KMP, no measurable change for Aho-Corasic. Looks like Aho-Corasic is hitting some hard CPU pipeline limit, caused by tight traversal of this 2 level cyclic memory access indirection (object -> array -> object). Currently trying to resolve this issue, with promising results. Once done, will update with full table of results (want to avoid excess noise here).